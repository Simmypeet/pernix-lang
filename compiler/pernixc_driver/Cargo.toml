[package]
name = "pernixc_driver"
version.workspace = true
authors.workspace = true
description.workspace = true
edition.workspace = true
rust-version.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
clap = { workspace = true }
pernixc_source_file = { path = "../pernixc_source_file" }
pernixc_diagnostic = { path = "../pernixc_diagnostic" }
pernixc_handler = { path = "../pernixc_handler" }
pernixc_lexical = { path = "../pernixc_lexical" }
pernixc_syntax = { path = "../pernixc_syntax" }
pernixc_semantic = { path = "../pernixc_semantic" }
rayon = { workspace = true }
serde_json = { workspace = true }
ron = { workspace = true }
bincode = { workspace = true }
serde_yaml = { workspace = true }
serde = { workspace = true }

[lints]
workspace = true
