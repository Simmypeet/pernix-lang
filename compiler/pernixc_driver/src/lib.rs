//! Contains the main `run()` function for the compiler.

use std::{fs::File, io::BufWriter, path::PathBuf, process::ExitCode};

use argument::Arguments;
use clap::Args;
use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    term::termcolor::WriteColor,
};
use dashmap::DashMap;
use pernixc_diagnostic::Report;
use pernixc_lexical::tree::RelativeLocation;
use pernixc_source_file::{
    ByteIndex, GlobalSourceID, Location, SourceFile, SourceMap,
};
use pernixc_target::TargetID;
use postcard::ser_flavors::io::WriteFlavor;
use term::get_coonfig;

pub mod argument;
pub mod term;

struct ReportTerm<'a> {
    config: codespan_reporting::term::Config,
    err_writer: &'a mut dyn WriteColor,
}

impl ReportTerm<'_> {
    fn report(
        &mut self,
        source_map: &mut SourceMap,
        diagnostic: &Diagnostic<GlobalSourceID>,
    ) {
        let _ = codespan_reporting::term::emit(
            self.err_writer,
            &self.config,
            source_map,
            diagnostic,
        );
    }
}

fn pernix_diagnostic_to_codespan_diagnostic(
    diagostic: pernixc_diagnostic::Diagnostic<ByteIndex>,
) -> codespan_reporting::diagnostic::Diagnostic<GlobalSourceID> {
    let mut result = match diagostic.severity {
        pernixc_diagnostic::Severity::Error => {
            codespan_reporting::diagnostic::Diagnostic::error()
        }
        pernixc_diagnostic::Severity::Warning => {
            codespan_reporting::diagnostic::Diagnostic::warning()
        }
        pernixc_diagnostic::Severity::Info => {
            codespan_reporting::diagnostic::Diagnostic::note()
        }
    }
    .with_message(diagostic.message);

    if let Some((span, label)) = diagostic.span {
        result = result.with_labels(
            std::iter::once({
                let mut primary = Label::primary(span.source_id, span.range());

                if let Some(label_message) = label {
                    primary = primary.with_message(label_message);
                }

                primary
            })
            .chain(diagostic.related.into_iter().map(|x| {
                Label::secondary(x.span.source_id, x.span.range())
                    .with_message(x.message)
            }))
            .collect(),
        );
    }

    if let Some(msg) = diagostic.help_message {
        result.with_notes(vec![msg])
    } else {
        result
    }
}

fn rel_pernix_diagnostic_to_codespan_diagnostic(
    diagostic: pernixc_diagnostic::Diagnostic<RelativeLocation>,
    source_map: &SourceMap,
    token_trees_by_source_id: &DashMap<
        GlobalSourceID,
        pernixc_lexical::tree::Tree,
    >,
) -> codespan_reporting::diagnostic::Diagnostic<GlobalSourceID> {
    let mut result = match diagostic.severity {
        pernixc_diagnostic::Severity::Error => {
            codespan_reporting::diagnostic::Diagnostic::error()
        }
        pernixc_diagnostic::Severity::Warning => {
            codespan_reporting::diagnostic::Diagnostic::warning()
        }
        pernixc_diagnostic::Severity::Info => {
            codespan_reporting::diagnostic::Diagnostic::note()
        }
    }
    .with_message(diagostic.message);

    if let Some((span, label)) = diagostic.span {
        result = result.with_labels(
            std::iter::once({
                let token_tree = token_trees_by_source_id
                    .get(&span.source_id)
                    .expect("Source ID not found in token trees");

                let source_file = source_map.get(span.source_id).unwrap();
                let begin =
                    span.start.to_absolute_index(&source_file, &token_tree);
                let end = span.end.to_absolute_index(&source_file, &token_tree);

                let mut primary = Label::primary(span.source_id, begin..end);

                if let Some(label_message) = label {
                    primary = primary.with_message(label_message);
                }

                primary
            })
            .chain(diagostic.related.into_iter().map(|x| {
                let token_tree = token_trees_by_source_id
                    .get(&x.span.source_id)
                    .expect("Source ID not found in token trees");

                let source_file = source_map.get(x.span.source_id).unwrap();

                let begin =
                    x.span.start.to_absolute_index(&source_file, &token_tree);
                let end =
                    x.span.end.to_absolute_index(&source_file, &token_tree);

                Label::secondary(x.span.source_id, begin..end)
                    .with_message(x.message)
            }))
            .collect(),
        );
    }

    if let Some(msg) = diagostic.help_message {
        result.with_notes(vec![msg])
    } else {
        result
    }
}

fn create_root_source_file(
    argument: &Arguments,
    source_map: &mut SourceMap,
    report_term: &mut ReportTerm,
) -> Option<GlobalSourceID> {
    let file = match File::open(&argument.command.input().file) {
        Ok(file) => file,
        Err(error) => {
            let msg = Diagnostic::error().with_message(format!(
                "{}: {error}",
                argument.command.input().file.display()
            ));
            report_term.report(source_map, &msg);

            return None;
        }
    };

    let source_file =
        match SourceFile::load(file, argument.command.input().file.clone()) {
            Ok(file) => file,
            Err(pernixc_source_file::Error::Io(error)) => {
                let msg = Diagnostic::error().with_message(format!(
                    "{}: {error}",
                    argument.command.input().file.display()
                ));
                report_term.report(source_map, &msg);

                return None;
            }
            Err(pernixc_source_file::Error::Utf8(error)) => {
                let msg = Diagnostic::error().with_message(format!(
                    "{}: {error}",
                    argument.command.input().file.display()
                ));
                report_term.report(source_map, &msg);

                return None;
            }
        };

    let id = source_map.register(TargetID::Local, source_file);

    Some(TargetID::Local.make_global(id))
}

/// The input to the compiler.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Args)]
pub struct Input {
    /// The input file to compile.
    ///
    /// This file is the root source file of the compilation; the module will
    /// stem from this file.
    pub file: PathBuf,

    /// The name of the target; if not specified, the target name will be
    /// inferred from the file name.
    #[clap(short = 't', long = "target")]
    pub target_name: Option<String>,

    /// The paths to the `plib` library to link to the target.
    #[clap(short = 'l', long = "link")]
    pub library_paths: Vec<PathBuf>,

    /// The path to the incremental compilation data.
    #[clap(long = "inc")]
    pub incremental_path: Option<PathBuf>,

    /// Whether to show the progress of the compilation.
    #[clap(long)]
    pub show_progress: bool,
}

/// Runs the program with the given arguments.
#[must_use]
#[allow(clippy::too_many_lines)]
pub fn run(
    argument: &Arguments,
    err_writer: &mut dyn WriteColor,
    _out_writer: &mut dyn WriteColor,
) -> ExitCode {
    let mut report_term = ReportTerm { config: get_coonfig(), err_writer };
    let mut source_map = SourceMap::new();

    let Some(root_source_id) =
        create_root_source_file(argument, &mut source_map, &mut report_term)
    else {
        return ExitCode::FAILURE;
    };

    let mut query_runtime = pernixc_query::runtime::Runtime::default();
    let mut serde = pernixc_query::serde::Serde::default();

    pernixc_query_base::register_runtime(&mut query_runtime, &mut serde);

    let target_name =
        argument.command.input().target_name.clone().unwrap_or_else(|| {
            source_map
                .get(root_source_id)
                .unwrap()
                .full_path()
                .file_stem()
                .unwrap()
                .to_string_lossy()
                .to_string()
        });

    let database = match pernixc_query_base::start_query_database(
        &mut source_map,
        root_source_id,
        argument.command.input().library_paths.iter().map(PathBuf::as_path),
        &target_name,
        argument
            .command
            .input()
            .incremental_path
            .as_ref()
            .map(|x| (x.as_path(), &serde)),
    ) {
        Ok(database) => database,
        Err(error) => {
            let msg = match error {
                pernixc_query_base::Error::OpenIncrementalFileIO(error) => {
                    Diagnostic::error().with_message(format!(
                        "Failed to open incremental file: {error}"
                    ))
                }
                pernixc_query_base::Error::IncrementalFileDeserialize(
                    error_kind,
                ) => Diagnostic::error().with_message(format!(
                    "Failed to load (deserialize) incremental file: \
                     {error_kind}"
                )),
                pernixc_query_base::Error::ReadIncrementalFileIO(error) => {
                    Diagnostic::error().with_message(format!(
                        "Failed to read incremental file: {error}"
                    ))
                }
            };

            report_term.report(&mut source_map, &msg);
            return ExitCode::FAILURE;
        }
    };

    for diagnostic in database.module_parsing_errors {
        let codespan_reporting = match diagnostic {
            pernixc_query_base::module::Error::Lexical(error) => {
                pernix_diagnostic_to_codespan_diagnostic(
                    error.report(&source_map),
                )
            }

            pernixc_query_base::module::Error::Syntax(error) => {
                pernix_diagnostic_to_codespan_diagnostic(
                    error.report(
                        &database
                            .token_trees_by_source_id
                            .get(&error.source_id)
                            .unwrap(),
                    ),
                )
            }

            pernixc_query_base::module::Error::RootSubmoduleConflict(
                root_submodule_conflict,
            ) => rel_pernix_diagnostic_to_codespan_diagnostic(
                root_submodule_conflict.report(()),
                &source_map,
                &database.token_trees_by_source_id,
            ),

            pernixc_query_base::module::Error::SourceFileLoadFail(
                source_file_load_fail,
            ) => rel_pernix_diagnostic_to_codespan_diagnostic(
                source_file_load_fail.report(()),
                &source_map,
                &database.token_trees_by_source_id,
            ),
            pernixc_query_base::module::Error::ModuleRedefinition(
                module_redefinition,
            ) => rel_pernix_diagnostic_to_codespan_diagnostic(
                module_redefinition.report(()),
                &source_map,
                &database.token_trees_by_source_id,
            ),
        };

        report_term.report(&mut source_map, &codespan_reporting);
    }

    if let Some(incremental_path) =
        argument.command.input().incremental_path.as_ref()
    {
        let incremental_file = match File::create(incremental_path) {
            Ok(file) => file,
            Err(error) => {
                let msg = Diagnostic::error().with_message(format!(
                    "Failed to create incremental file: {error}"
                ));
                report_term.report(&mut source_map, &msg);
                return ExitCode::FAILURE;
            }
        };
        let serializable_database = database.database.serializable(&serde);

        let writer = BufWriter::new(incremental_file);
        let flavor = WriteFlavor::new(writer);

        // Serialize using postcard and write to file
        if let Err(error) =
            postcard::serialize_with_flavor(&serializable_database, flavor)
        {
            let msg = Diagnostic::error().with_message(format!(
                "Failed to serialize incremental file: {error}"
            ));
            report_term.report(&mut source_map, &msg);
            return ExitCode::FAILURE;
        }
    }

    ExitCode::SUCCESS
}
