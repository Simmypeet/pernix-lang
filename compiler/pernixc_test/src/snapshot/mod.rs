#![allow(missing_docs)]

use std::{
    io::BufRead,
    path::{Path, PathBuf},
};

use clap::Parser;
use insta::assert_snapshot;
use pernixc_driver::{
    argument::{Arguments, Check, Command},
    Input,
};

#[test]
#[allow(clippy::manual_assert)]
fn main() {
    let scandir =
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("src").join("snapshot");

    std::env::set_current_dir(&scandir).unwrap();

    let mut passed = Vec::new();
    let mut failed = Vec::new();
    visit_dirs(&scandir, &mut |file_path: &Path| {
        let relative_to_scandir = file_path.strip_prefix(&scandir).unwrap();
        let result = std::panic::catch_unwind(|| {
            // for each file that has the `main.pnx` name, run the test
            test(relative_to_scandir);
        });

        if result.is_err() {
            failed.push(relative_to_scandir.to_path_buf());
        } else {
            passed.push(relative_to_scandir.to_path_buf());
        }
    })
    .unwrap();

    println!("Passed tests: {}", passed.len());
    for file in &passed {
        println!("  - {}", file.display());
    }

    if failed.is_empty() {
        println!("All tests passed!");
    } else {
        println!("Failed tests: {}", failed.len());
        for file in &failed {
            println!("  - {}", file.display());
        }
        panic!("Some tests failed, see the output above for details.");
    }
}

// one possible implementation of walking a directory only visiting files
fn visit_dirs(dir: &Path, cb: &mut dyn FnMut(&Path)) -> std::io::Result<()> {
    if dir.is_dir() {
        for entry in std::fs::read_dir(dir)? {
            let entry = entry?;
            let path = entry.path();
            if path.is_dir() {
                visit_dirs(&path, cb)?;
            } else if let Some(path) = entry.file_name().to_str() {
                if path == "main.pnx" {
                    cb(&entry.path());
                }
            }
        }
    }
    Ok(())
}

fn test(file_path: &Path) {
    let mut err_writer =
        codespan_reporting::term::termcolor::NoColor::new(Vec::new());
    let mut out_writer =
        codespan_reporting::term::termcolor::NoColor::new(Vec::new());

    let file = std::fs::File::open(file_path).unwrap();
    let buf_reader = std::io::BufReader::new(file);

    // read for the custom cli interface starting with `##` comment
    let arguments = buf_reader.lines().next().map(|x| x.unwrap()).map_or_else(
        || Arguments {
            command: Command::Check(Check {
                input: Input {
                    file: file_path.to_path_buf(),
                    target_name: None,
                    library_paths: Vec::new(),
                    incremental_path: None,
                    show_progress: false,
                },
            }),
        },
        |mut first_line| {
            if first_line.starts_with("##") {
                // replace the `{}` with the file path (if any) first_line =
                first_line = first_line
                    .replace("{}", file_path.to_string_lossy().as_ref());

                first_line.strip_prefix("##").unwrap();

                Arguments::parse_from(first_line.split_whitespace())
            } else {
                Arguments {
                    command: Command::Check(Check {
                        input: Input {
                            file: file_path.to_path_buf(),
                            target_name: None,
                            library_paths: Vec::new(),
                            incremental_path: None,
                            show_progress: false,
                        },
                    }),
                }
            }
        },
    );

    let _ = pernixc_driver::run(&arguments, &mut err_writer, &mut out_writer);

    let stderr_string = String::from_utf8(err_writer.into_inner()).unwrap();
    let stout_string = String::from_utf8(out_writer.into_inner()).unwrap();

    let mut settings = insta::Settings::clone_current();

    // Convert windows paths to Unix Paths.
    settings.add_filter(r"\\\\?([\w\d.])", "/$1");
    // Convert crlf to lf.
    settings.add_filter(r"\r\n", "\n");
    // Replace the IO error message with a generic one.
    settings.add_filter(
        r": [^(\r\n]*\(os error (\d+)\)",
        ": general IO error (os error $1)",
    );

    settings.set_snapshot_path(file_path.parent().unwrap());
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_snapshot_suffix();
    let _guard = settings.bind_to_scope();

    let assert_string =
        format!("stderr:\n{stderr_string}\n\nstdout:\n{stout_string}");

    assert_snapshot!("snapshot", assert_string);
}
