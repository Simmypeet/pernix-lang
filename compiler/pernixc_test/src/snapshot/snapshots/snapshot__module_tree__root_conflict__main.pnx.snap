---
source: compiler/pernixc_test/src/snapshot/mod.rs
expression: assert_string
---
stderr:
error: the submodule of the root source file ends up pointing to the root source file itself
  ╭─ module_tree/root_conflict/main.pnx:3:15
  │
3 │ public module main
  │               ^^^^ this ends up loading the file `module_tree/root_conflict/main.pnx`, which is the current module itself
  │
  = consider renaming the submodule name to avoid this conflict



stdout:
